File: .claude/skills/web-content-fetcher/scripts/extract_article.js
#!/usr/bin/env node
/**
 * Article Content Extractor using Mozilla Readability
 *
 * Usage:
 *   node extract_article.js <html_file>
 *
 * Output:
 *   Clean article content in markdown format to stdout
 *
 * Requirements:
 *   npm install @mozilla/readability jsdom
 */

const fs = require('fs');

let Readability, JSDOM;

try {
  ({ Readability } = require('@mozilla/readability'));
  ({ JSDOM } = require('jsdom'));
} catch (error) {
  console.error('Error: Required libraries not installed.');
  console.error(`Details: ${error.message}`);
  console.error('Install with: npm install @mozilla/readability jsdom');
  process.exit(1);
}

/**
 * Convert HTML to markdown-like text format
 */
function htmlToMarkdown(html) {
  const dom = new JSDOM(html);
  const doc = dom.window.document;
  const result = [];

  function processNode(node, depth = 0) {
    if (node.nodeType === 3) {
      // Text node
      const text = node.textContent.trim();
      if (text) {
        result.push(text);
      }
      return;
    }

    if (node.nodeType !== 1) {
      return;
    } // Not an element

    const tagName = node.tagName.toLowerCase();

    switch (tagName) {
      case 'h1':
        result.push(`\n# ${node.textContent.trim()}\n`);
        break;
      case 'h2':
        result.push(`\n## ${node.textContent.trim()}\n`);
        break;
      case 'h3':
        result.push(`\n### ${node.textContent.trim()}\n`);
        break;
      case 'h4':
        result.push(`\n#### ${node.textContent.trim()}\n`);
        break;
      case 'h5':
        result.push(`\n##### ${node.textContent.trim()}\n`);
        break;
      case 'h6':
        result.push(`\n###### ${node.textContent.trim()}\n`);
        break;
      case 'p':
        result.push(`\n${node.textContent.trim()}\n`);
        break;
      case 'br':
        result.push('\n');
        break;
      case 'a':
        const href = node.getAttribute('href');
        const text = node.textContent.trim();
        if (href && text) {
          result.push(`[${text}](${href})`);
        } else {
          result.push(text);
        }
        break;
      case 'strong':
      case 'b':
        result.push(`**${node.textContent.trim()}**`);
        break;
      case 'em':
      case 'i':
        result.push(`*${node.textContent.trim()}*`);
        break;
      case 'code':
        result.push(`\`${node.textContent.trim()}\``);
        break;
      case 'blockquote':
        result.push(`\n> ${node.textContent.trim()}\n`);
        break;
      case 'ul':
      case 'ol':
        result.push('\n');
        Array.from(node.children).forEach((li) => {
          if (li.tagName.toLowerCase() === 'li') {
            result.push(`- ${li.textContent.trim()}\n`);
          }
        });
        result.push('\n');
        break;
      default:
        // Process children
        Array.from(node.childNodes).forEach((child) =>
          processNode(child, depth + 1)
        );
    }
  }

  processNode(doc.body);
  return result
    .join(' ')
    .replace(/ +/g, ' ')
    .replace(/\n\n\n+/g, '\n\n')
    .trim();
}

/**
 * Extract article from HTML file
 */
function extractArticle(htmlFilePath) {
  // Read HTML file
  if (!fs.existsSync(htmlFilePath)) {
    throw new Error(`File not found: ${htmlFilePath}`);
  }

  const htmlContent = fs.readFileSync(htmlFilePath, 'utf-8');

  // Parse with JSDOM
  const dom = new JSDOM(htmlContent);
  const reader = new Readability(dom.window.document);
  const article = reader.parse();

  if (!article) {
    throw new Error('Failed to extract article content');
  }

  // Convert to markdown format
  const markdown = htmlToMarkdown(article.content);

  return `# ${article.title}\n\n${markdown}`;
}

// Main execution
function main() {
  if (process.argv.length !== 3) {
    console.error('Usage: node extract_article.js <html_file>');
    process.exit(1);
  }

  const htmlFile = process.argv[2];

  try {
    const articleContent = extractArticle(htmlFile);
    console.log(articleContent);
  } catch (error) {
    console.error(`Error extracting article: ${error.message}`);
    process.exit(1);
  }
}

main();

File: .claude/skills/web-content-fetcher/scripts/extract_article.py
#!/usr/bin/env python3
"""
Article Content Extractor using BeautifulSoup and Readability

Usage:
    python extract_article.py <html_file>

Output:
    Clean article content in markdown format to stdout

Requirements:
    pip install beautifulsoup4 lxml readability-lxml
"""

import sys
from pathlib import Path

try:
    from bs4 import BeautifulSoup
    from readability import Document
except ImportError:
    print("Error: Required libraries not installed.", file=sys.stderr)
    print("Install with: pip install beautifulsoup4 lxml readability-lxml", file=sys.stderr)
    sys.exit(1)


def extract_article(html_file_path):
    """
    Extract main article content from HTML file.

    Args:
        html_file_path: Path to HTML file

    Returns:
        Cleaned article content in markdown format
    """
    # Read HTML file
    html_path = Path(html_file_path)
    if not html_path.exists():
        raise FileNotFoundError(f"File not found: {html_file_path}")

    with open(html_path, 'r', encoding='utf-8') as f:
        html_content = f.read()

    # Use readability to extract main content
    doc = Document(html_content)
    title = doc.title()
    summary_html = doc.summary()

    # Parse with BeautifulSoup for cleaning
    soup = BeautifulSoup(summary_html, 'lxml')

    # Remove script and style tags
    for tag in soup(['script', 'style', 'nav', 'footer', 'aside']):
        tag.decompose()

    # Extract text with structure preservation
    result = [f"# {title}\n"]

    for element in soup.find_all(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p', 'ul', 'ol', 'blockquote']):
        tag_name = element.name
        text = element.get_text(strip=True)

        if not text:
            continue

        if tag_name == 'h1':
            result.append(f"\n# {text}\n")
        elif tag_name == 'h2':
            result.append(f"\n## {text}\n")
        elif tag_name == 'h3':
            result.append(f"\n### {text}\n")
        elif tag_name == 'h4':
            result.append(f"\n#### {text}\n")
        elif tag_name == 'h5':
            result.append(f"\n##### {text}\n")
        elif tag_name == 'h6':
            result.append(f"\n###### {text}\n")
        elif tag_name == 'p':
            result.append(f"{text}\n")
        elif tag_name in ['ul', 'ol']:
            for li in element.find_all('li', recursive=False):
                li_text = li.get_text(strip=True)
                result.append(f"- {li_text}\n")
            result.append("\n")
        elif tag_name == 'blockquote':
            result.append(f"> {text}\n")

    return '\n'.join(result)


def main():
    if len(sys.argv) != 2:
        print("Usage: python extract_article.py <html_file>", file=sys.stderr)
        sys.exit(1)

    html_file = sys.argv[1]

    try:
        article_content = extract_article(html_file)
        print(article_content)
    except Exception as e:
        print(f"Error extracting article: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
